src/librustc/middle/region.rs

```diff
@@ -1142,6 +1143,68 @@ fn region_maps<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId)
     Rc::new(maps)
 }
 
+struct YieldFinder(Option<Span>);
+
+impl<'tcx> Visitor<'tcx> for YieldFinder {
+    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'tcx> {
+        NestedVisitorMap::None
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx hir::Expr) {
+        if let hir::ExprYield(..) = expr.node {
+            if self.0.is_none() {
+                self.0 = Some(expr.span);
+            }
+        }
+
+        intravisit::walk_expr(self, expr);
+    }
+}
+
+impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
+    /// Checks whether the given code extent contains a `yield`. If so,
+    /// returns `Some(span)` with the span of the "first" yield we find.
+    pub fn yield_in_extent(self, extent: CodeExtent) -> Option<Span> {
```

Comment:
This is `O(n^2)` in large functions. I think it would be better to use a tree search to create a set/hashmap of all extents that contain a yield.